File: api/app/routers/presets.py
"""Preset-API: Pipeline-Presets verwalten und anwenden."""

import asyncio

from fastapi import APIRouter, HTTPException

from app.schemas.preset import (
    PresetApplyRequest,
    PresetApplyResponse,
    PresetCreate,
    PresetResponse,
    PresetStep,
    PresetUpdate,
)
from app.services import preset_service

router = APIRouter(prefix="/presets", tags=["presets"])


def _to_response(data: dict) -> PresetResponse:
    return PresetResponse(
        id=data["id"],
        name=data["name"],
        description=data.get("description", ""),
        created_at=data["created_at"],
        updated_at=data["updated_at"],
        steps=[PresetStep(**s) for s in data.get("steps", [])],
    )


@router.get("", response_model=list[PresetResponse])
async def list_presets():
    """Alle Presets auflisten."""
    items = await asyncio.to_thread(preset_service.list_presets)
    return [_to_response(p) for p in items]


@router.post("", response_model=PresetResponse)
async def create_preset(body: PresetCreate):
    """Neues Preset erstellen."""
    steps = [s.model_dump() for s in body.steps]
    data = await asyncio.to_thread(
        preset_service.create_preset,
        body.name,
        body.description,
        steps,
    )
    return _to_response(data)


@router.get("/{preset_id}", response_model=PresetResponse)
async def get_preset(preset_id: str):
    """Preset laden."""
    data = await asyncio.to_thread(preset_service.get_preset, preset_id)
    if not data:
        raise HTTPException(404, detail="Preset nicht gefunden")
    return _to_response(data)


@router.put("/{preset_id}", response_model=PresetResponse)
async def update_preset(preset_id: str, body: PresetUpdate):
    """Preset aktualisieren."""
    steps = [s.model_dump() for s in body.steps] if body.steps is not None else None
    data = await asyncio.to_thread(
        preset_service.update_preset,
        preset_id,
        name=body.name,
        description=body.description,
        steps=steps,
    )
    if not data:
        raise HTTPException(404, detail="Preset nicht gefunden")
    return _to_response(data)


@router.delete("/{preset_id}", status_code=204)
async def delete_preset(preset_id: str):
    """Preset löschen."""
    deleted = await asyncio.to_thread(preset_service.delete_preset, preset_id)
    if not deleted:
        raise HTTPException(404, detail="Preset nicht gefunden")


@router.post("/{preset_id}/apply", response_model=PresetApplyResponse)
async def apply_preset(preset_id: str, body: PresetApplyRequest):
    """
    Preset auf Asset anwenden — gibt Execution-Plan zurück.
    Führt keine Steps aus; Nutzer bestätigt und führt manuell aus.
    """
    try:
        plan, applicable, skipped = await asyncio.to_thread(
            preset_service.compute_execution_plan,
            preset_id,
            body.asset_id,
            body.start_from_step,
        )
    except FileNotFoundError as e:
        raise HTTPException(404, detail=str(e)) from e

    preset = await asyncio.to_thread(preset_service.get_preset, preset_id)
    steps_total = len(preset.get("steps", [])) if preset else 0

    return PresetApplyResponse(
        preset_id=preset_id,
        asset_id=body.asset_id,
        steps_total=steps_total,
        steps_applicable=applicable,
        steps_skipped=skipped,
        execution_plan=plan,
    )

File: frontend/src/api/presets.ts
import { apiClient } from "./client.js";
import type { ProviderParamValue } from "./generation.js";

export interface PresetStep {
  step: string;
  provider: string | null;
  params: Record<string, ProviderParamValue>;
}

export interface Preset {
  id: string;
  name: string;
  description: string;
  created_at: string;
  updated_at: string;
  steps: PresetStep[];
}

export interface PresetCreate {
  name: string;
  description: string;
  steps: PresetStep[];
}

export interface PresetApplyRequest {
  asset_id: string;
  start_from_step?: number;
  dry_run?: boolean;
}

export interface ExecutionPlanItem {
  step_index: number;
  step: string;
  provider: string | null;
  params: Record<string, ProviderParamValue>;
  status: "skipped" | "applicable";
  reason: string | null;
}

export interface PresetApplyResponse {
  preset_id: string;
  asset_id: string;
  steps_total: number;
  steps_applicable: number;
  steps_skipped: number;
  execution_plan: ExecutionPlanItem[];
}

export interface PresetSuggestions {
  suggested_name: string;
  steps: PresetStep[];
}

export async function listPresets(): Promise<Preset[]> {
  const { data } = await apiClient.get<Preset[]>("/presets");
  return data;
}

export async function getPreset(presetId: string): Promise<Preset> {
  const { data } = await apiClient.get<Preset>(`/presets/${presetId}`);
  return data;
}

export async function createPreset(
  preset: PresetCreate
): Promise<Preset> {
  const { data } = await apiClient.post<Preset>("/presets", preset);
  return data;
}

export async function updatePreset(
  presetId: string,
  updates: Partial<PresetCreate>
): Promise<Preset> {
  const { data } = await apiClient.put<Preset>(`/presets/${presetId}`, updates);
  return data;
}

export async function deletePreset(presetId: string): Promise<void> {
  await apiClient.delete(`/presets/${presetId}`);
}

export async function applyPreset(
  presetId: string,
  req: PresetApplyRequest
): Promise<PresetApplyResponse> {
  const { data } = await apiClient.post<PresetApplyResponse>(
    `/presets/${presetId}/apply`,
    req
  );
  return data;
}

export async function getPresetSuggestions(
  assetId: string
): Promise<PresetSuggestions> {
  const { data } = await apiClient.get<PresetSuggestions>(
    `/assets/${assetId}/preset-suggestions`
  );
  return data;
}
